### ISSUE TITLE
ratelimit incoming requests

### ISSUE THREAD

**Issue by vtcao297** [OP-TEE/optee_os#3936]
I see that with this commit (https://github.com/OP-TEE/optee_os/commit/27b5e34b47fbd3028586ce81d62ce46c55432667) there is a mechanism to retrieve the number of threads which OPTEE can handle simultaneously.  My question is does the optee client kernel driver currently use this information to do any type of rate limiting logic to prevent DOS types of attacks?

**Comment by etienne-lms** [comment]
No. This new service is used by hypervisors. Refer to the link @lorc put in the commit message. Current OP-TEE Linux driver does not use this information.
I am not sure limiting the invocation rate will help DoS attacks but there is maybe something to do in this area.

**Comment by vtcao297** [comment]
If the driver doesn't do any limiting, does optee core itself do any limit based the number of threads?  What happens to incoming requests if all owned threads are busy?  Is some there kind of queue or will optee return an error on that request?

**Comment by etienne-lms** [comment]
Yes.

In OP-TEE Core, at entry, `thread_handle_std_smc()` calls `thread_alloc_and_run()` to find a TEE thread. If no available thread is found, the later function returns and [`thread_handle_std_smc()`](https://github.com/OP-TEE/optee_os/blob/3.9.0/core/arch/arm/kernel/thread_optee_smc.c#L75) returns to non-secure Linux driver with return value `OPTEE_SMC_RETURN_ETHREAD_LIMIT`.

In OP-TEE linux driver, on invocation return value `OPTEE_SMC_RETURN_ETHREAD_LIMIT`, the Linux thread is [queued](https://elixir.bootlin.com/linux/v5.7/source/drivers/tee/optee/call.c#L149) to wait for a free TEE thread. It is woken when another thread exists OP-TEE Core, maining that, maybe, a TEE thread has been free.

In conclusion: if all TEE threads are busy, an invocation will wait for a free thread.
